source/fuzz/transformation_changing_memory_semantics.cpp

```diff
@@ -0,0 +1,350 @@
+            ->GetDef(needed_atomic_instruction->GetSingleWordInOperand(
+                message_.memory_semantics_operand_index() == 0 ? 2 : 3))
+            ->GetSingleWordInOperand(0);
+  }
+  auto first_5bits_new_memory_semantics = static_cast<SpvMemorySemanticsMask>(
+      new_memory_sematics_value & kMemorySemanticsLowerBitmask);
+  auto first_5bits_old_memory_semantics = static_cast<SpvMemorySemanticsMask>(
+      old_memory_sematics_value & kMemorySemanticsLowerBitmask);
+  auto memory_model = static_cast<SpvMemoryModel>(
+      ir_context->module()->GetMemoryModel()->GetSingleWordInOperand(1));
+
+  if (!IsValidConversion(needed_atomic_instruction->opcode(),
+                         first_5bits_old_memory_semantics,
+                         first_5bits_new_memory_semantics, memory_model)) {
+    return false;
+  }
+
+  // The higher bits value of old and new memory semantics id must be equal.
+  auto higher_bits_new_memory_semantics = static_cast<SpvMemorySemanticsMask>(
+      new_memory_sematics_value & kMemorySemanticsHigherBitmask);
+  auto higher_bits_old_memory_semantics = static_cast<SpvMemorySemanticsMask>(
+      old_memory_sematics_value & kMemorySemanticsHigherBitmask);
+  if (higher_bits_new_memory_semantics != higher_bits_old_memory_semantics) {
+    return false;
+  }
+
+  // Instructions that take two memory semantics and id needed to change are
+  // unequal, the equal id must be stronger than unequal id. Unequal id can't be
+  // released or acquire/release memory semantics.
+  if ((needed_atomic_instruction->opcode() == SpvOpAtomicCompareExchange ||
+       needed_atomic_instruction->opcode() == SpvOpAtomicCompareExchangeWeak) &&
+      message_.memory_semantics_operand_index() == 1) {
+    auto equal_id_memory_semantics_first_5bits =
+        static_cast<SpvMemorySemanticsMask>(
+            ir_context->get_def_use_mgr()
+                ->GetDef(needed_atomic_instruction->GetSingleWordInOperand(2))
+                ->GetSingleWordInOperand(0) &
+            kMemorySemanticsLowerBitmask);
+
+    if (first_5bits_new_memory_semantics >
+        equal_id_memory_semantics_first_5bits) {
+      return false;
+    }
+    if (first_5bits_new_memory_semantics == SpvMemorySemanticsReleaseMask ||
+        first_5bits_new_memory_semantics ==
+            SpvMemorySemanticsAcquireReleaseMask) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+void TransformationChangingMemorySemantics::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  auto needed_atomic_instruction =
+      FindInstruction(message_.atomic_instruction(), ir_context);
+
+  uint32_t needed_index =
+      GetNeededIndex(needed_atomic_instruction->opcode(),
+                     message_.memory_semantics_operand_index());
+
+  needed_atomic_instruction->SetInOperand(
+      needed_index, {message_.memory_semantics_new_value_id()});
+}
+
+uint32_t TransformationChangingMemorySemantics::GetNeededIndex(
```

Comment:
I suggest using a better name, like `GetMemorySemanticsOperandIndex(SpvOp opcode, uint32_t zero_or_one)`.

Under `SpvOpMemoryBarrier`, assert that `zero_or_one` is 0. List all other atomic instructions as cases. Under the default case, assert(false).

Reply:
Done.